[CodeCompletion] Give up fast-completion if dependent files are modified

rintaro opened the pull request:
Check if dependencies are modified since the last checking.
Dependencies:

 - Other source files in the current module
 - Dependent files collected by the dependency tracker

When:

 - If the last dependency check was over N (defaults to 5) seconds ago

Invalidate if:

 - The dependency file is missing
 - The modification time of the dependecy is greater than the last check
 - If the modification time is zero, compare the contents using the file system from the previous completion and the current completion (in-memory filesystem)

rdar://problem/62336432

rintaro committed:
[CodeCompletion] Give up fast-completion if dependent files are modified

Check if dependencies are modified since the last checking.
Dependencies:

 - Other source files in the current module
 - Dependent files collected by the dependency tracker

When:

 - If the last dependency check was over N (defaults to 5) seconds ago

Invalidate if:

 - The dependency file is missing
 - The modification time of the dependecy is greater than the last check
 - If the modification time is zero, compare the content using the file
   system from the previous completion and the current completion

rdar://problem/62336432

rintaro commented:
@swift-ci Please build toolchain macOS

rintaro commented:
@swift-ci Please build toolchain macOS

rintaro commented:
@swift-ci Please build toolchain macOS

rintaro commented:
@swift-ci Please build toolchain macOS

rintaro commented:
@swift-ci Please smoke test

rintaro committed:
[CodeCompletion] Added check dependency test for complete.open

rintaro committed:
[Tests] Mark check dependency tests cases 'REQUIRES: shell'

These test cases use shell variables etc.

rintaro commented:
@swift-ci Please test

rintaro commented:
@swift-ci Please test

akyrtzi commented on include/swift/IDE/CompletionInstance.h:
```diff
@@ -38,20 +39,29 @@ makeCodeCompletionMemoryBuffer(const llvm::MemoryBuffer *origBuf,
 
   std::unique_ptr<CompilerInstance> CachedCI;
   llvm::hash_code CachedArgHash;
+  llvm::sys::TimePoint<> CacheUsedTimestamp;
```
What is this used for, is it leftover?

rintaro replied:
Yes, it's left over.

rintaro commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -162,10 +165,79 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
+    // If the last modification time is zero, this file is probably from a
+    // virtual file system. We need to check the content.
+    if (lastModTime == llvm::sys::TimePoint<>()) {
+      if (&CI.getFileSystem() == &FS)
```
> Are these ever going to be equal, or in practice we'll always get a new VFS object?

I don't think so. I believe this is unreachable.

> Even if the pointers are equal, is it correct to assume the contents did not change? The VFS object could be mutable.

You are right. We perhaps should retain all buffers of the VFS dependencies. But I think it's good enough for now because the VFS objects should be always different.

benlangmuir replied:
I would say keep a hash not the whole buffer.

rintaro replied:
OK. I updated the PR to keep hashes of the dependencies.

rintaro commented on test/SourceKit/CodeComplete/complete_checkdeps_bridged.swift:
```diff
@@ -0,0 +1,74 @@
+  /*HERE*/
+}
+
+// FIXME: Windows CI doesn't support bash array syntax i.e. 'ARRY=( .... )'
```
I didn't know `shell`. Thank you!

rintaro commented:
@swift-ci Please test

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -162,10 +165,77 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
+
+    // Ignore the current file and synthesized files.
+    if (filename.empty() || filename.front() == '<' ||
+        filename.equals(currentFileName))
```
Comparing filename seems potentially brittle - shouldn't we know which file is the completion buffer directly without looking at the name?

rintaro replied:
Updated the PR.

rintaro committed:
[CodeCompletion] Tweak fast completion dependency checking

- Detect same file with bufferID instead of the file name
- Compare virtual in-memory filesystem content with hash value

rintaro commented:
@swift-ci Please test

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -162,10 +165,120 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
+        // Stop processing if the modification time is not zero. Probably this
+        // is not a virtual file system.
+        if (stat->getLastModificationTime() != llvm::sys::TimePoint<>())
+          return true;
```
I think I'm not understanding something: why do we want to stop as soon as we see the first "normal file"? Why can't we have virtual mixed in with regular?

rintaro replied:
I just want to save the `stat` time if the filesystem is not in-memory.
AFAIU, in reality, we don't mix in-memory file with normal file.

benlangmuir replied:
The filesystem you get from our builtin FileSystemProvider in sourcekitd is an overlay, so if a call is not handled by the InMemoryFilesystem, it will fall through to the RealFilesystem.

rintaro replied:
Ah, thank you for notifying! Updated to just ignore the non-zero modification time here.

benlangmuir commented on lib/IDE/CompletionInstance.cpp:
```diff
@@ -162,10 +165,120 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
+/// For each dependency files in CI, run \p callback until the callback returns
+/// \c true. Returns \c true if any callback call returns \c true, \c false
+/// otherwise.
+bool forEachDependenciesUntilTrue(
```
Should be singular dependency in the name: `forEachDependencyUntilTrue`

benlangmuir commented on test/SourceKit/CodeComplete/complete_checkdeps_bridged.swift:
```diff
@@ -0,0 +1,74 @@
+// RUN: %empty-directory(%t/Frameworks)
+// RUN: %empty-directory(%t/MyProject)
+
+// RUN: COMPILER_ARGS=( \
```
I'm not sure how supported it is to assume you can set shell variables and have them persist across run lines.  This may work in practice, but are we sure lit will never run commands in a fresh environment?

rintaro replied:
I'm not 100% certain this is officially supported. But several test cases rely on the behavior e.g. https://github.com/apple/swift/blob/master/test/ModuleInterface/ModuleCache/module-cache-deployment-target-irrelevant.swift.

benlangmuir replied:
Okay, maybe we can follow up with people who know lit well, but no need to block this PR.

rintaro commented:
@swift-ci Please test

rintaro commented:
@swift-ci Please test
